hail/python/hailtop/utils/utils.py

```diff
@@ -158,8 +170,54 @@ def __init__(self, worker_pool):
         self._waiting = True
         if self._n_complete == self._n_submitted:
             self._done.set()
+        if not self._ignore_errors and self._errors:
+            self._done.set()
+
         await self._done.wait()
 
+        if self._errors:
+            raise self._errors[0]
+
+
+class WaitableBunch:
+    def __init__(self, shared_pool):
+        self.shared_pool = shared_pool
+        self._errors = []
+        self._done = asyncio.Event()
+        self._waiting = False
+        self._n_submitted = 0
+        self._n_complete = 0
+
+    async def call(self, f, *args, **kwargs):
+        assert not self._waiting
+        self._n_submitted += 1
+
+        async def invoke():
+            try:
+                await f(*args, **kwargs)
+            except Exception as err:  # pylint: disable=broad-except
+                self._errors.append(err)
+                self._done.set()
+            finally:
+                self._n_complete += 1
+                if self._waiting and (self._n_complete == self._n_submitted):
+                    self._done.set()
+
+        await self.shared_pool._worker_pool.call(invoke)
+
+    async def wait(self):
+        assert not self._waiting
+        self._waiting = True
+        if self._n_complete == self._n_submitted:
+            self._done.set()
+        if self._errors:
+            self._done.set()
+
+        await self._done.wait()
+
+        if self._errors:
+            raise self._errors[0]
```

Comment:
I'm inclined to raise a new error that at least includes all the names of the errors, in the odd case that there are many different kinds. Still good to `... from self._errors[0]`

Reply:
I don't think this is necessary.